controllers/image_generator.go

```diff
@@ -0,0 +1,359 @@
+*/
+
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/go-logr/logr"
+	configv1 "github.com/openshift/api/config/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+/*
+The image generator builds and deletes pod VM images for a cloud provider. It uses Kubernetes jobs to do this.
+Make sure
+Ensure that all job YAML files meet the following guidelines:
+1. Job files should be located at config/peerpods/
+2. all file names must follow this format: <cloud-provider>-[CVM|VM]-image-[create|delete]-job.yaml
+3. parallelism: 1
+4. completions: 1 (or none)
+5. backoffLimit: 1
+6. restartPolicy: Never
+7. job and pod names describes cloud-provider and operation
+8. Get configuration values from peer-pods-secret and peer-pods-cm volumes and predefined enviroment variables
+9. make sure create/delete jobs are based on the same configuration sources
+10. utilize precreated podvm binaries conatiner if possible
+11. all jobs must abort on image creation failure
+
+Every image creation job must obey to the following guidelines:
+1. create the podvm image and all other required cloud resouces needed for it
+2. Create a container named result that outputs the image ID (and nothing but the image ID) once the image is ready.
+3. image ID can be shared between containers using emptyDir volume
+
+Every image deletion job should preform the following:
+1. based on the exect same configurtion creation job is using
+2. deletes the image and all created resouces
+*/
+
+const (
+	peerpodsCMName          = "peer-pods-cm"
+	peerpodsCMAWSImageKey   = "PODVM_AMI_ID"
+	peerpodsCMAzureImageKey = "AZURE_IMAGE_ID"
+	defaultVMType           = "VM"
+)
+
+type ImageGenerator struct {
+	client    client.Client        // controller-runtime client
+	clientset kubernetes.Interface // k8s clientset
+	Log       logr.Logger
+
+	provider     string
+	CMimageIDKey string
+}
+
+func NewImageGenerator(client client.Client) *ImageGenerator {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil
+	}
+
+	return &ImageGenerator{
+		client:       client,
+		clientset:    clientset,
+		provider:     "", // set by getCloudProvider
+		CMimageIDKey: "", // set by getCloudProvider
+		Log:          ctrl.Log.WithName("image-generator"),
+	}
+}
+
+func (r *ImageGenerator) getCloudProviderFromInfra() string {
+	// TODO: first check if it's indeed openshift
+	infrastructure := &configv1.Infrastructure{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infrastructure)
```

Comment:
Where does this "cluster" namespace comes from ?

Reply:
it's the namespace of the cluster's "infrastructure" CR, it gets the cluster's region from there